File: library/src/main/java/org/xmtp/android/library/codecs/ContentTypeId.kt
package org.xmtp.android.library.codecs

import uniffi.xmtpv3.FfiContentTypeId

typealias ContentTypeId = org.xmtp.proto.message.contents.Content.ContentTypeId

class ContentTypeIdBuilder {
    companion object {
        fun builderFromAuthorityId(
            authorityId: String,
            typeId: String,
            versionMajor: Int,
            versionMinor: Int,
        ): ContentTypeId =
            ContentTypeId
                .newBuilder()
                .also {
                    it.authorityId = authorityId
                    it.typeId = typeId
                    it.versionMajor = versionMajor
                    it.versionMinor = versionMinor
                }.build()

        fun fromFfi(ffiContentTypeId: FfiContentTypeId): ContentTypeId =
            ContentTypeId
                .newBuilder()
                .also {
                    it.authorityId = ffiContentTypeId.authorityId
                    it.typeId = ffiContentTypeId.typeId
                    it.versionMajor = ffiContentTypeId.versionMajor.toInt()
                    it.versionMinor = ffiContentTypeId.versionMinor.toInt()
                }.build()
    }
}

val ContentTypeId.id: String
    get() = "$authorityId:$typeId:$versionMajor.$versionMinor"

val ContentTypeId.description: String
    get() = "$authorityId/$typeId:$versionMajor.$versionMinor"

File: Sources/XMTPiOS/Codecs/ContentTypeID.swift
//
//  ContentTypeID.swift
//
//
//  Created by Pat Nakajima on 11/28/22.
//

public typealias ContentTypeID = Xmtp_MessageContents_ContentTypeId

public extension ContentTypeID {
	init(authorityID: String, typeID: String, versionMajor: Int, versionMinor: Int) {
		self.init()
		self.authorityID = authorityID
		self.typeID = typeID
		self.versionMajor = UInt32(versionMajor)
		self.versionMinor = UInt32(versionMinor)
	}
}

public extension ContentTypeID {
	var id: String {
		"\(authorityID):\(typeID):\(versionMajor).\(versionMinor)"
	}

	var description: String {
		"\(authorityID)/\(typeID):\(versionMajor).\(versionMinor)"
	}
}
